File: src/main/java/fit/nlu/tmdt/modules/favorite/service/FavoriteService.java
package fit.nlu.tmdt.modules.favorite.service;

import fit.nlu.tmdt.modules.favorite.dto.response.FavoriteResponse;

import java.util.List;

/**
 * Favorite Service Interface
 */
public interface FavoriteService {

    /**
     * Get user's favorites
     */
    List<FavoriteResponse> getUserFavorites(Long userId);

    /**
     * Add room to favorites
     */
    FavoriteResponse addFavorite(Long userId, Long roomId);

    /**
     * Remove room from favorites
     */
    void removeFavorite(Long userId, Long roomId);

    /**
     * Check if room is in favorites
     */
    boolean isFavorite(Long userId, Long roomId);

    /**
     * Get favorite count for user
     */
    long getFavoriteCount(Long userId);

    /**
     * Check multiple rooms for favorites
     */
    List<Long> checkFavoriteRooms(Long userId, List<Long> roomIds);

    /**
     * Get suggested rooms for user
     */
    List<FavoriteResponse> getSuggestedRooms(Long userId, int limit);

    /**
     * Get most viewed rooms
     */
    List<FavoriteResponse> getMostViewedRooms(int limit);

    /**
     * Get latest rooms
     */
    List<FavoriteResponse> getLatestRooms(int limit);
}

File: frontend/src/services/favoriteService.ts
import api from './api';
import type { ApiResponse, PostResponse } from '@/types';

export interface FavoriteResponse {
  id: number;
  roomId: number;
  roomTitle: string;
  roomAddress: string;
  roomImageUrl: string;
  roomPrice: number;
  priceType: 'MONTHLY' | 'QUARTERLY' | 'YEARLY';
  landlordId: number;
  landlordName: string;
  landlordAvatar: string;
  addedAt: string;
}

export const favoriteService = {
  /**
   * Get user's favorites
   */
  getUserFavorites: async (): Promise<FavoriteResponse[]> => {
    const response = await api.get<ApiResponse<FavoriteResponse[]>>('/v1/favorites');
    return response.data.data!;
  },

  /**
   * Add room to favorites
   */
  addFavorite: async (roomId: number | string): Promise<FavoriteResponse> => {
    const response = await api.post<ApiResponse<FavoriteResponse>>(`/v1/favorites/${roomId}`);
    return response.data.data!;
  },

  /**
   * Remove room from favorites
   */
  removeFavorite: async (roomId: number | string): Promise<void> => {
    await api.delete(`/v1/favorites/${roomId}`);
  },

  /**
   * Check if room is favorite
   */
  isFavorite: async (roomId: number | string): Promise<boolean> => {
    const response = await api.get<ApiResponse<{ isFavorite: boolean }>>(`/v1/favorites/check/${roomId}`);
    return response.data.data?.isFavorite || false;
  },
};

export default favoriteService;
